### DISCUSSION TITLE
JWT Token not found

### DISCUSSION THREAD

**Discussion by salamisodikiolawale** [lexik/LexikJWTAuthenticationBundle#1023] [General]
Hello, I don't really understand what is going on. I have been encountering this error for 3 days. 
I get my token but when I put a firewall rule on a route and I send the request with the obtained token I get the error 401 Token not found.

This is my security file :

`security:
    enable_authenticator_manager: true
    # https://symfony.com/doc/current/security.html#registering-the-user-hashing-passwords
    password_hashers:
        Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: 'auto'
        App\Entity\User:
            algorithm: auto

    # https://symfony.com/doc/current/security.html#loading-the-user-the-user-provider
    providers:
        # used to reload user from session & other features (e.g. switch_user)
        app_user_provider:
            entity:
                class: App\Entity\User
                property: email
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false

        login:
            pattern: ^/api/login
            stateless: true
            json_login:
                check_path: /api/login_check # or api_login_check as defined in config/routes.yaml
                username_path: email
                password_path: password
                success_handler: lexik_jwt_authentication.handler.authentication_success
                failure_handler: lexik_jwt_authentication.handler.authentication_failure
        
        article:
            pattern:   ^/api/articles
            methods: [POST, PUT, PATCH, DELETE]
            stateless: true
            guard:
                authenticators:
                    - lexik_jwt_authentication.jwt_token_authenticator
        api:
            pattern:   ^/api
            stateless: true
            guard:
                authenticators:
                    - lexik_jwt_authentication.jwt_token_authenticator
        

        main:
            lazy: true
            provider: app_user_provider

            # activate different ways to authenticate
            # https://symfony.com/doc/current/security.html#the-firewall

            # https://symfony.com/doc/current/security/impersonating_user.html
            # switch_user: true

    # Easy way to control access for large sections of your site
    # Note: Only the *first* access control that matches will be used
    access_control:
        # - { path: ^/admin, roles: ROLE_ADMIN }
        # - { path: ^/profile, roles: ROLE_USER }
        - { path: ^/api/login, roles: PUBLIC_ACCESS }
        - { path: ^/api,       roles: IS_AUTHENTICATED_FULLY }
        - { path: ^/api/articles, roles: IS_AUTHENTICATED_FULLY, methods: [POST, PUT, PATCH, DELETE] }

when@test:
    security:
        password_hashers:
            # By default, password hashers are resource intensive and take time. This is
            # important to generate secure password hashes. In tests however, secure hashes
            # are not important, waste resources and increase test times. The following
            # reduces the work factor to the lowest possible values.
            Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface:
                algorithm: auto
                cost: 4 # Lowest possible value for bcrypt
                time_cost: 3 # Lowest possible value for argon
                memory_cost: 10 # Lowest possible value for argon

`

My route file:

`
api_login_check:
  path: /api/login_check
`

**Comment by fd6130** [comment]
Hi, just replace your `guard` to `jwt` instead (guard is deprecated):

```yaml
article:
        pattern:   ^/api/articles
        methods: [POST, PUT, PATCH, DELETE]
        stateless: true
        jwt: ~
    api:
        pattern:   ^/api
        stateless: true
        jwt: ~
```

**Reply by salamisodikiolawale** [reply]
Hello, after to test my code with your suggestion (jwt) i get the same error.
{
    "code": 401,
    "message": "JWT Token not found"
}`
`

**Reply by fd6130** [reply]
> Hello, after to test my code with your suggestion (jwt) i get the same error. { "code": 401, "message": "JWT Token not found" }` `

Can you show me your request header? Or how do you call the api?

**Reply by salamisodikiolawale** [reply]
Actully, i use postman.
I active Bearer token 
Header  postman : Content-type : application/json , Accept : application/json
settings: JSON
body:
`
{
    "name": "Article 110",
    "price": 183,
    "quantity": 71,
    "createAt": "2022-05-13T21:52:30+00:00",
    "type": "LENTILLE_DE_COULEUR",
    "gender": "FEMME",
    "image": "img02.jpg",
    "make": "/api/make_articles/36"
}
`

**Reply by salamisodikiolawale** [reply]
i will use my app angular after.

**Reply by salamisodikiolawale** [reply]
I would like to be sure, can the generation of the priv and publ key pair by openssl be a sub problem option?

**Reply by chalasr** [reply]
It's not a key issue: the error you get means that no token has been passed to the request. Try using curl from the command line as shown in the readme. Also make sure to check this if your webserver is Apache: https://github.com/lexik/LexikJWTAuthenticationBundle/blob/2.x/Resources/doc/index.rst#important-note-for-apache-users

**Reply by salamisodikiolawale** [reply]
I'm using apache as a server and I looked at the doc but I still get the error  

`
curl  -H "Authorization: Bearer {token here}"  http://127.0.0.1:8081/api/invoices

<VirtualHost *:80>
    DocumentRoot /var/www/public
    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    <Directory /var/www/public>
        AllowOverride None
        Order Allow,Deny
        Allow from All
 
        <IfModule mod_rewrite.c>
            RewriteEngine On
            RewriteCond %{REQUEST_FILENAME} !-f
            RewriteRule ^(.*)$ index.php [QSA,L]
        </IfModule>
    </Directory>
</VirtualHost>

`

**Reply by fd6130** [reply]
Are you using docker? If yes, can you try non-docker approach? Like install database, apache into your local machine and try?

**Reply by salamisodikiolawale** [reply]
Yes i using docker. i have some bug in local environnement so i need to work in container.

**Reply by fd6130** [reply]
My guessing is that something wrong with the docker (maybe), if non-docker approach works then which mean the docker configuration got issue .

**Reply by Spomky** [reply]
> Actully, i use postman.
> I active Bearer token

What is the value pass to this header? It should be the token only (without the scheme "Bearer")

**Reply by salamisodikiolawale** [reply]
> 

Hello, i had tested in my local environnement without docker so API it run with JWT with postman.
But i don't get solution with docker.

**Reply by salamisodikiolawale** [reply]
But i need to continuous with docker. 
With Docker i had tested many solutiion but i get same error with docker (JWT not found).
If you have a proposal don't hesitate.

**Reply by codebysaadbouh** [reply]
Hello @salamisodikiolawale, I finally found the problem. Actually there are libraries that only work in HTTPS and I think [LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle) is one of them, so it is necessary to activate the TLS with the command ``symfony server:ca:install`` (https://symfony.com/doc/current/setup/symfony_server.html#enabling-tls)

**Reply by salamisodikiolawale** [reply]
> Hello @salamisodikiolawale, I finally found the problem. Actually there are libraries that only work in HTTPS and I think [LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle) is one of them, so it is necessary to activate the TLS with the command `symfony server:ca:install` (https://symfony.com/doc/current/setup/symfony_server.html#enabling-tls)

With Docker environnement ? 
Because without Docker i had not need to https with lexikjwt.

**Reply by codebysaadbouh** [reply]
but without docker when you do a ``symfony server:start`` you don't have https:... ?

**Reply by salamisodikiolawale** [reply]
> but without docker when you do a `symfony server:start` you don't have https:... ?

My origin problem is jwt, not protocol https. 
without docker my api works fine with lexik bundle token management but once I containerize my api with docker that's when I have the token problem.

**Reply by salamisodikiolawale** [reply]
or maybe I didn't understand your solution.
